### PR TITLE
[SPARK-39976][SQL] ArrayIntersect should handle null in left expression correctly

### PR DESCRIPTION
**Pull request by AngersZhuuuu**
### What changes were proposed in this pull request?
`ArrayInterscet` miss judge if null contains in right expression's hash set.

```
>>> a = [1, 2, 3]
>>> b = [3, None, 5]
>>> df = spark.sparkContext.parallelize(data).toDF(["a","b"])
>>> df.show()
+---------+------------+
|        a|           b|
+---------+------------+
|[1, 2, 3]|[3, null, 5]|
+---------+------------+

>>> df.selectExpr("array_intersect(a,b)").show()
+---------------------+
|array_intersect(a, b)|
+---------------------+
|                  [3]|
+---------------------+

>>> df.selectExpr("array_intersect(b,a)").show()
+---------------------+
|array_intersect(b, a)|
+---------------------+
|            [3, null]|
+---------------------+
```

In origin code gen's code path, when handle `ArrayIntersect`'s array1, it use the below code
```
        def withArray1NullAssignment(body: String) =
          if (left.dataType.asInstanceOf[ArrayType].containsNull) {
            if (right.dataType.asInstanceOf[ArrayType].containsNull) {
              s"""
                 |if ($array1.isNullAt($i)) {
                 |  if ($foundNullElement) {
                 |    $nullElementIndex = $size;
                 |    $foundNullElement = false;
                 |    $size++;
                 |    $builder.$$plus$$eq($nullValueHolder);
                 |  }
                 |} else {
                 |  $body
                 |}
               """.stripMargin
            } else {
              s"""
                 |if (!$array1.isNullAt($i)) {
                 |  $body
                 |}
               """.stripMargin
            }
          } else {
            body
          }
```
We have a flag `foundNullElement` to indicate if array2 really contains a null value. But when implement https://issues.apache.org/jira/browse/SPARK-36829, misunderstand the meaning of `ArrayType.containsNull`,
so when implement  `SQLOpenHashSet.withNullCheckCode()`
```
  def withNullCheckCode(
      arrayContainsNull: Boolean,
      setContainsNull: Boolean,
      array: String,
      index: String,
      hashSet: String,
      handleNotNull: (String, String) => String,
      handleNull: String): String = {
    if (arrayContainsNull) {
      if (setContainsNull) {
        s"""
           |if ($array.isNullAt($index)) {
           |  if (!$hashSet.containsNull()) {
           |    $hashSet.addNull();
           |    $handleNull
           |  }
           |} else {
           |  ${handleNotNull(array, index)}
           |}
         """.stripMargin
      } else {
        s"""
           |if (!$array.isNullAt($index)) {
           | ${handleNotNull(array, index)}
           |}
         """.stripMargin
      }
    } else {
      handleNotNull(array, index)
    }
  }
```
The code path of `  if (arrayContainsNull && setContainsNull) `  is misinterpreted that array's openHashSet really have a null value. 

In this pr we add a new parameter `additionalCondition ` to complements the previous implementation of `foundNullElement`. Also refactor the method's parameter name.

### Why are the changes needed?
Fix data correct issue

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Added UT

### CODE REVIEW

**Comment by AngersZhuuuu** [comment]
ping @cloud-fan @tgravescs

**Comment by tgravescs** [comment]
@HyukjinKwon

**Comment by AngersZhuuuu** [comment]
Looks like UT failed not caused by this pr

File: sql/catalyst/src/main/scala/org/apache/spark/sql/util/SQLOpenHashSet.scala [hunk 1]
```diff
@@ -85,12 +85,13 @@ object SQLOpenHashSet {
...
       handleNotNull: (String, String) => String,
-      handleNull: String): String = {
+      handleNull: String,
+      nullConditions: Seq[String] = Seq.empty): String = {
     if (arrayContainsNull) {
       if (setContainsNull) {
         s"""
            |if ($array.isNullAt($index)) {
-           |  if (!$hashSet.containsNull()) {
+           |  if (!$hashSet.containsNull()${nullConditions.map(" &&" + _).mkString("")}) {
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/util/SQLOpenHashSet.scala:94]:
```suggestion
           |  if (!$hashSet.containsNull()${nullConditions.map(" && " + _).mkString)}) {
```

Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/util/SQLOpenHashSet.scala:94]: Why is the implementation different from interpreted path? e.g., `SQLOpenHashSet.withNullCheckFunc`

**Comment by HyukjinKwon** [comment]
cc @rednaxelafx FYI

File: sql/catalyst/src/main/scala/org/apache/spark/sql/util/SQLOpenHashSet.scala (hunk 1 above)
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/util/SQLOpenHashSet.scala:94]: same question

Comment (AngersZhuuuu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/util/SQLOpenHashSet.scala]: Caused by `ArrayType`'s `containsNull` is not reliable.  
When use `Literal.create(Seq(1d), ArrayType(DoubleType))` this array's `containsNull` is also true

Comment (AngersZhuuuu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/util/SQLOpenHashSet.scala]: The whole stage codegen's code seems wrong ....in origins code it's ok seems it use a flag foundNullElement to flag if array2 contains null.

```
        def withArray1NullAssignment(body: String) =
          if (left.dataType.asInstanceOf[ArrayType].containsNull) {
            if (right.dataType.asInstanceOf[ArrayType].containsNull) {
              s"""
                 |if ($array1.isNullAt($i)) {
                 |  if ($foundNullElement) {
                 |    $nullElementIndex = $size;
                 |    $foundNullElement = false;
                 |    $size++;
                 |    $builder.$$plus$$eq($nullValueHolder);
                 |  }
                 |} else {
                 |  $body
                 |}
               """.stripMargin
            } else {
              s"""
                 |if (!$array1.isNullAt($i)) {
                 |  $body
                 |}
               """.stripMargin
            }
          } else {
            body
          }
```

Comment (AngersZhuuuu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/util/SQLOpenHashSet.scala]: How about current's code path?

**Comment by AngersZhuuuu** [comment]
> [#37436 (comment)](https://github.com/apache/spark/pull/37436#discussion_r941255226) You did not answer the question. Why it's only an issue with codegen?

Update it in the pr desc

File: sql/catalyst/src/main/scala/org/apache/spark/sql/util/SQLOpenHashSet.scala (hunk 1 above)
Comment (AngersZhuuuu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/util/SQLOpenHashSet.scala]: > Why is the implementation different from interpreted path? e.g., `SQLOpenHashSet.withNullCheckFunc`

Their own code generation methods are different. There will be containsNull related judgments in codegen, so the extraction methods are different. The problem this time is caused by the surface problem of the wrong understanding of the field.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala [hunk 2]
```diff
@@ -4158,7 +4158,8 @@ case class ArrayIntersect(left: Expression, right: Expression) extends ArrayBina
              |$nullElementIndex = $size;
              |$size++;
              |$builder.$$plus$$eq($nullValueHolder);
-           """.stripMargin)
+           """.stripMargin,
+          Some(s"$hashSet.containsNull()"))
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:4162]: The interpreted code path is:
```
          val withArray1NullCheckFunc = SQLOpenHashSet.withNullCheckFunc(elementType, hsResult,
            (value: Any) => withArray1NaNCheckFunc(value),
            () =>
              if (hs.containsNull()) {
                arrayBuffer += null
              }
          )
```
Can we follow it in the codegen path? e.g. `handleNull` should check `if (hashSet.containsNull())`

Comment (AngersZhuuuu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala]: Updated

**Comment by AngersZhuuuu** [comment]
Any more suggestion?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala [hunk 3]
```diff
@@ -4155,9 +4155,11 @@ case class ArrayIntersect(left: Expression, right: Expression) extends ArrayBina
           right.dataType.asInstanceOf[ArrayType].containsNull,
           array1, i, hashSetResult, withArray1NaNCheckCodeGenerator,
           s"""
-             |$nullElementIndex = $size;
-             |$size++;
-             |$builder.$$plus$$eq($nullValueHolder);
+             |if ($hashSet.containsNull()) {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:4158]: does ArrayExcept have the bug?

Comment (AngersZhuuuu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:4158]: No

**Comment by cloud-fan** [comment]
thanks, merging to master/3.3!

**Comment by cloud-fan** [comment]
@AngersZhuuuu can you create 2 more PRs for 3.2 and 3.1? thanks!

**Comment by AngersZhuuuu** [comment]
> @AngersZhuuuu can you create 2 more PRs for 3.2 and 3.1? thanks!

Sure

**Comment by dongjoon-hyun** [comment]
Hi, @cloud-fan .
Could you update Apache JIRA, https://issues.apache.org/jira/browse/SPARK-39976 ?

**Comment by AngersZhuuuu** [comment]
> @AngersZhuuuu can you create 2 more PRs for 3.2 and 3.1? thanks!

After check, 3.2 and 3.1 don't have this problem, this problem is caused by the refactor pr

**Comment by cloud-fan** [comment]
@AngersZhuuuu do you know which refactor PR caused this bug?

**Comment by AngersZhuuuu** [comment]
> @AngersZhuuuu do you know which refactor PR caused this bug?

https://issues.apache.org/jira/browse/SPARK-36829

**Pull request closed without merging**